stream_alert/rule_processor/threat_intel.py

```diff
@@ -17,15 +17,372 @@
+                    norm_types2ioc_types_mapping[norm_type] = ioc_type
+
+                sub_norm_types[norm_type] = orig_types
+            norm_types_mapping[log_src] = sub_norm_types
+
+        # Class variable stores mapping between CEF normalized types and IOC types
+        cls.__normalized_types2ioc_types = norm_types2ioc_types_mapping
+        # Class variable stores Data Normalization types mapping.
+        cls.__normalized_types = norm_types_mapping
+
+    @staticmethod
+    def _validate_type_mapping(mapping_str):
+        """Static method to extract normalized type and IOC type from qualified str
+
+        Args:
+            mapping_str (str): A qualified string has pattern 'normalized_type:ioc_type'
+
+        Returns:
+            (bool): First return indicate if the string a qualifited string contains
+                both normalized CEF type and IOC type.
+            (str): Second return is normalized type.
+            (str): Last return is IOC type.
+        """
+        normalized_type = None
+        ioc_type = None
+
+        splitted_str = mapping_str.split(':')
+        if len(splitted_str) == 1:
+            normalized_type = splitted_str[0]
+        elif len(splitted_str) == 2:
+            normalized_type = splitted_str[0]
+            ioc_type = splitted_str[1].split('_')[-1]
+        else:
+            LOGGER.info('Key %s in conf/types.json is incorrect', mapping_str)
+            return False, None, None
+
+        if normalized_type and ioc_type:
+            return True, normalized_type, ioc_type
+
+        return False, normalized_type, None
+
+    def _process_ioc(self, ioc_collections):
+        """Check if any info is malicious by querying DynamoDB IOC table
+
+        Args:
+            ioc_collections (list): A list of StreamIoc instances.
+        """
+        # Segment data before calling DynamoDB table with batch_get_item.
+        for subset in StreamThreatIntel._segment(ioc_collections):
+            query_values = [ioc.value for ioc in subset]
+            query_result = self._query(query_values)
+            for value in ioc_collections:
+                for ioc in query_result:
+                    if value.value == ioc[PRIMARY_KEY]:
+                        value.itype = ioc[SUB_TYPE_KEY]
+                        value.is_ioc = True
+                        continue
+
+    @staticmethod
+    def _segment(ioc_collections):
+        """Static method to segment ioc_collections in to smaller set.
+        Batch query to dynamodb supports up to 100 items.
+
+        Args:
+            ioc_collections (list): A list of StreamIoc instances
+
+        Returns:
+            (list): List of subset of StreamIoc instances
+        """
+        result = []
+        end = len(ioc_collections)
+        for index in range(0, end, MAX_QUERY_CNT):
+            result.append(ioc_collections[index:min(index+MAX_QUERY_CNT, end)])
+        return result
+
+    exceptions_to_backoff = (ClientError)
```

Comment:
this could be a constant in the top of the file

Reply:
Unfortunately, I can not make it as a constant. Pylint complaint
```
Starting pylint script
************* Module stream_alert.rule_processor.threat_intel
C: 36, 0: Invalid class name "EXCEPTIONS_TO_BACKOFF" (invalid-name)
```
After digging pylint's error, looks pylint is able to recognize `ClientError` is a class, so `EXCEPTIONS_TO_BACKOFF` will be a class. The [regular expression](http://pylint-messages.wikidot.com/messages:c0103) of class name is different from constant. 
```
Constant: (([A-Z_][A-Z0-9_]*)|(__.*__))$
Class: [A-Z_][a-zA-Z0-9]+$
```
Translate to human readable version is that Class variable is started with Capital letter or underscore, but no underscore in the middle. 
I don't think it is necessary to make change to this part.

Reply:
@chunyong-lin this error is because you're accidentally just making `exceptions_to_backoff` reference `ClientError` when what you're really tryin to do is create a tuple with one item of `ClientError` within it. To accomplish this, add a trailing comma to this like so:

`exceptions_to_backoff = (ClientError,)`

or as a constant: `EXCEPTIONS_TO_BACKOFF = (ClientError,)`